src/fancy-transformers/fragment.ts

```diff
@@ -0,0 +1,177 @@
+import logging = require('../logging/logging');
+
+var log :logging.Log = new logging.Log('fancy-transformers');
+
+var takeByte_ = (buffer:ArrayBuffer) : number => {
+  var bytes=new Uint8Array(buffer);
+  return bytes[0];
+}
+
+var dropByte_ = (buffer:ArrayBuffer) : ArrayBuffer => {
+  var bytes=new Uint8Array(buffer)
+  var result = new Uint8Array(buffer.byteLength-1);
+  var fromIndex : number = 1;
+  var toIndex : number = 0;
+  while(toIndex < result.length) {
+    result[toIndex] = bytes[fromIndex];
+    toIndex=toIndex+1;
+    fromIndex=fromIndex+1;
+  }
+
+  return result.buffer;
+}
+
+var encodeByte_ = (num:number) : ArrayBuffer => {
+  var bytes = new Uint8Array(1);
+  bytes[0]=num;
+  return bytes.buffer;
+}
+
+var split_ = (buffer:ArrayBuffer, firstLen:number) : Array<ArrayBuffer> => {
+  var bytes=new Uint8Array(buffer)
+  var lastLen : number = buffer.byteLength-firstLen;
+  var first = new Uint8Array(firstLen);
+  var last = new Uint8Array(lastLen);
+  var fromIndex : number = 0;
+  var toIndex : number = 0;
+  while(toIndex < first.length) {
+    first[toIndex] = bytes[fromIndex];
+    toIndex=toIndex+1;
+    fromIndex=fromIndex+1;
+  }
+
+  toIndex=0;
+  while(toIndex < last.length) {
+    last[toIndex] = bytes[fromIndex];
+    toIndex=toIndex+1;
+    fromIndex=fromIndex+1;
+  }
+
+  return [first.buffer, last.buffer];
+}
+
+/* Takes a two byte (network byte order) representation of a number and returns
+ * the number.
+ */
+ // TODO(bwiley): Byte order may be backward
+ // TODO(bwiley): Fix type error
+var decodeLength_ = (buffer:ArrayBuffer) : number => {
+  var bytes = new Uint8Array(buffer);
+  var result = (bytes[0] << 8) | bytes[1];
+  return result;
+}
+
+
+class Fragment {
+  public length : number;
+  public id : ArrayBuffer;
+  public index : number;
+  public count : number;
+  public payload : ArrayBuffer;
+  public padding : ArrayBuffer;
+
+  public constructor(length:number, id:ArrayBuffer, index:number, count:number, payload:ArrayBuffer, padding:ArrayBuffer) {
+    this.length=length;
+    this.id=id;
+    this.index=index;
+    this.count=count;
+    this.payload=payload;
+    this.padding=padding;
+  }
+
+  static randomId = () : ArrayBuffer => {
+    var bytes = new Uint8Array(32);
+    for (var i = 0; i < bytes.byteLength; i++) {
+      bytes[i] = Math.floor(Math.random()*255);
+    }
+    return bytes.buffer;
+  }
+
+  static decodeFragment = (buffer:ArrayBuffer) : Fragment => {
+//    log.debug('Decode fragment %1 %2', buffer.byteLength, length);
+    var parts = split_(buffer, 2);
+    var lengthBytes = parts[0];
+    var length = decodeLength_(lengthBytes);
```

Comment:
These two variables: length and lengthBytes, don't tell me much about what they're the lengths of.